File: results/0.25/SQLITE3-DB/sqlite3-db-claude-3-5-haiku-20241022.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <hammer/hammer.h>

typedef struct {
    uint8_t magic[16];
    uint16_t page_size;
    uint8_t write_version;
    uint8_t read_version;
    uint8_t reserved_space;
    uint8_t max_embedded_payload;
    uint8_t min_embedded_payload;
    uint8_t leaf_payload_fraction;
    uint32_t file_change_counter;
    uint32_t database_size_pages;
    uint32_t first_freelist_page;
    uint32_t total_freelist_pages;
    uint32_t schema_cookie;
    uint32_t schema_format;
    uint32_t default_encoding;
    uint32_t user_version;
    uint32_t incremental_vacuum_mode;
    uint32_t application_id;
} SQLiteHeader;

HParser* parse_magic_header() {
    return h_token((const uint8_t*)"SQLite format 3\0", 16);
}

HParser* parse_page_size() {
    return h_int_range(h_uint16(), 512, 65536);
}

HParser* parse_version() {
    return h_choice(h_token((const uint8_t*)"\1", 1), h_token((const uint8_t*)"\2", 1), NULL);
}

HParser* parse_sqlite_header() {
    return h_sequence(
        parse_magic_header(),
        parse_page_size(),
        parse_version(),
        parse_version(),
        h_uint8(),
        h_uint8(),
        h_uint8(),
        h_uint8(),
        h_uint32(),
        h_uint32(),
        h_uint32(),
        h_uint32(),
        h_uint32(),
        h_uint32(),
        h_uint32(),
        h_uint32(),
        h_uint32(),
        h_uint32(),
        NULL
    );
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <sqlite_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation error");
        fclose(file);
        return 1;
    }

    if (fread(buffer, 1, file_size, file) != file_size) {
        perror("File read error");
        free(buffer);
        fclose(file);
        return 1;
    }
    fclose(file);

    HParser* parser = parse_sqlite_header();
    HParseResult* result = h_parse(parser, buffer, file_size);

    if (result && result->ast) {
        printf("SQLite file parsed successfully\n");
    } else {
        printf("Parsing failed\n");
    }

    h_parse_result_free(result);
    h_destroy_parser(parser);
    free(buffer);

    return 0;
}

File: results/0.25/SQLITE3-DB/sqlite3-db-claude-3-5-haiku-20241022.rs
use nom::{
    bytes::complete::{tag, take_while},
    combinator::{map, opt},
    multi::{count, many0},
    number::complete::{le_u16, le_u32, le_u64},
    sequence::{tuple, preceded},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct SQLiteHeader {
    magic: [u8; 16],
    page_size: u16,
    write_version: u8,
    read_version: u8,
    reserved_space: u8,
    max_payload_fraction: u8,
    min_payload_fraction: u8,
    leaf_payload_fraction: u8,
    file_change_counter: u32,
    database_size: u32,
    first_freelist_trunk_page: u32,
    total_freelist_pages: u32,
    schema_cookie: u32,
    schema_format: u32,
    default_cache_size: u32,
    largest_root_page: u32,
    text_encoding: u32,
    user_version: u32,
    incremental_vacuum_mode: u32,
    application_id: u32,
    reserved: [u8; 20],
    version_valid_for: u32,
    sqlite_version: u32,
}

fn parse_sqlite_header(input: &[u8]) -> IResult<&[u8], SQLiteHeader> {
    map(
        tuple((
            take_while(|_| true),
            le_u16,
            take_while(|_| true),
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            le_u32,
            take_while(|_| true),
            le_u32,
            le_u32,
        )),
        |(
            magic,
            page_size,
            _,
            file_change_counter,
            database_size,
            first_freelist_trunk_page,
            total_freelist_pages,
            schema_cookie,
            schema_format,
            default_cache_size,
            largest_root_page,
            text_encoding,
            user_version,
            incremental_vacuum_mode,
            application_id,
            _,
            version_valid_for,
            sqlite_version,
            reserved,
            write_version,
            read_version,
        )| SQLiteHeader {
            magic: magic.try_into().unwrap(),
            page_size,
            write_version: write_version as u8,
            read_version: read_version as u8,
            reserved_space: 0,
            max_payload_fraction: 64,
            min_payload_fraction: 32,
            leaf_payload_fraction: 32,
            file_change_counter,
            database_size,
            first_freelist_trunk_page,
            total_freelist_pages,
            schema_cookie,
            schema_format,
            default_cache_size,
            largest_root_page,
            text_encoding,
            user_version,
            incremental_vacuum_mode,
            application_id,
            reserved: reserved.try_into().unwrap(),
            version_valid_for,
            sqlite_version,
        },
    )(input)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <sqlite_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_sqlite_header(&buffer) {
        Ok((_, header)) => {
            println!("SQLite Database Header: {:?}", header);
            Ok(())
        }
        Err(e) => {
            eprintln!("Parsing error: {:?}", e);
            Err(e.into())
        }
    }
}